appender.go

```diff
@@ -132,149 +150,329 @@ func (a *Appender) AppendRow(args ...driver.Value) error {
+	}
+}
+
 // appendRowArray loads a row of values into the appender. The values are provided as an array.
 func (a *Appender) appendRowArray(args []driver.Value) error {
 	for i, v := range args {
+		columnInfo := a.columnInfos[i]
 		if v == nil {
-			if state := C.duckdb_append_null(*a.appender); state == C.DuckDBError {
-				return fmt.Errorf("couldn't append parameter %d", v)
-			}
+			setNull(&columnInfo, a.currentChunkSize)
 			continue
 		}
-
-		switch v := v.(type) {
-		case uint8:
-			set[uint8](a.chunkVectors[i], a.currentRow, v)
-		case int8:
-			set[int8](a.chunkVectors[i], a.currentRow, v)
-		case uint16:
-			set[uint16](a.chunkVectors[i], a.currentRow, v)
-		case int16:
-			set[int16](a.chunkVectors[i], a.currentRow, v)
-		case uint32:
-			set[uint32](a.chunkVectors[i], a.currentRow, v)
-		case int32:
-			set[int32](a.chunkVectors[i], a.currentRow, v)
-		case uint64:
-			set[uint64](a.chunkVectors[i], a.currentRow, v)
-		case int64:
-			set[int64](a.chunkVectors[i], a.currentRow, v)
-		case uint:
-			set[uint](a.chunkVectors[i], a.currentRow, v)
-		case int:
-			set[int](a.chunkVectors[i], a.currentRow, v)
-		case float32:
-			set[float32](a.chunkVectors[i], a.currentRow, v)
-		case float64:
-			set[float64](a.chunkVectors[i], a.currentRow, v)
-		case bool:
-			set[bool](a.chunkVectors[i], a.currentRow, v)
-		case []byte:
-			set[[]byte](a.chunkVectors[i], a.currentRow, v)
-		case UUID:
-			set[C.duckdb_hugeint](a.chunkVectors[i], a.currentRow, uuidToHugeInt(v))
-		case string:
-			str := C.CString(v)
-			C.duckdb_vector_assign_string_element(a.chunkVectors[i], C.uint64_t(a.currentRow), str)
-			C.free(unsafe.Pointer(str))
-		case time.Time:
-			var dt C.duckdb_timestamp
-			dt.micros = C.int64_t(v.UTC().UnixMicro())
-			set[C.duckdb_timestamp](a.chunkVectors[i], a.currentRow, dt)
-		default:
-			return fmt.Errorf("couldn't append unsupported parameter %d (type %T)", i, v)
-		}
+		columnInfo.function(a, &columnInfo, a.currentChunkSize, v)
 	}
+	a.currentChunkSize++
 
-	a.currentRow++
 	return nil
 }
 
 func (a *Appender) appendChunks() error {
 	// set the size of the current chunk to the current row
-	C.duckdb_data_chunk_set_size(a.chunks[a.currentChunkIdx], C.uint64_t(a.currentRow))
+	C.duckdb_data_chunk_set_size(a.chunks[len(a.chunks)-1], C.idx_t(a.currentChunkSize))
 
 	// append all chunks to the appender and destroy them
 	var state C.duckdb_state
 	for i, chunk := range a.chunks {
 		state = C.duckdb_append_data_chunk(*a.appender, chunk)
 		if state == C.DuckDBError {
 			dbErr := C.GoString(C.duckdb_appender_error(*a.appender))
-			return fmt.Errorf("duckdb error appending chunk %d of %d: %s", i+1, a.currentChunkIdx+1, dbErr)
+			return fmt.Errorf("duckdb error appending chunk %d of %d: %s", i+1, len(a.chunks), dbErr)
 		}
 		C.duckdb_destroy_data_chunk(&chunk)
 	}
+
+	// free the column types
+	for i := range a.columnInfos {
+		C.duckdb_destroy_logical_type(&a.columnTypesPtr[i])
+	}
+
+	C.free(unsafe.Pointer(a.columnTypes))
```

Comment:
I think that this has to go to `Close`, no? As of now, we also call this when we call `Flush`. However, from my understanding of `Flush`, we should still be able to use the `Appender` afterward, i.e., append more chunks. Can we also have a test for this? The test appends 1-2 chunks, calls `Flush`, appends one more chunk, and then calls `Close`. Or am I misunderstanding `Flush`?

Reply:
Yeah, we already do have a test for this.  In `appender_test.go`, `Flush` is called after appending 3000 rows.  But I think you're right, destroy/free should be in `Close` otherwise the logical types have to be initialized every time `Flush` is called.